File: Carelink.py
# Install the Python Requests library:
# `pip install requests`

import time
import sys
import requests
from lxml import html
from urlparse import urlparse
from datetime import datetime, timedelta

def startSession():
    return requests.session()


user_agent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/601.3.9 (KHTML, like Gecko)     Version/9.0.2 Safari/601.3.9"

def download(session, url):
    try:
        response = session.get(
            url=url,
            headers={
            },
        )
        return response.content
        # print('Response HTTP Status Code: {status_code}'.format(
            # status_code=response.status_code))
        # print('Response HTTP Response Body: {content}'.format(
            # content=response.content))
    except requests.exceptions.RequestException:
        # print('HTTP Request failed')
        return None


def request_all_reports(session):
    # All Reports
    # POST https://carelink.minimed.com/patient/main/requestProReport.do

    # TODO(nl): time zone
    startDate = datetime.now().replace(hour=0,minute=0,second=0,microsecond=0)
    delta = timedelta(-7)
    startDate = startDate + delta
    startDateString = startDate.strftime("%Y-%m-%dT%H:%M:%S-04:00")

    endDate = datetime.now().replace(hour=0,minute=0,second=0,microsecond=0)
    endDateString =  endDate.strftime("%Y-%m-%dT%H:%M:%S-04:00")

    clientTime = datetime.now()
    clientTimeString = clientTime.strftime("%m/%d/%y %I:M:%S %p")

    daysSelected = "" 
    d = datetime.now().replace(hour=0,minute=0,second=0,microsecond=0)
    for i in range(1, 7):
        delta = timedelta(-1*i)
        daysSelected += (d + delta).strftime("%Y-%m-%dT%H:%M:%S-04:00")

    data = {
                "masterDuration": "2",
                "startDateString": startDateString,
                "endDateString": endDateString,
                "clientTimeString": clientTimeString,
                "selected[0]": "on",
                "selected[3]": "on",
                "customerID": "553090",
                "dailyDetails": "1,2,3",
                "selected[1]": "on",
                "selected[4]": "on",
                "dailySelected": endDateString,
                "selected[2]": "on",
                "selected[5]": "on",
            }
            
    try:
        response = session.post(
            url="https://carelink.minimed.com/patient/main/requestProReport.do",
            headers={
                "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
            },
            data=data,
        )
        page = html.fromstring(response.content)
        reportURL = page.xpath('//*[@id="content"]/form/@action')
        if len(reportURL) == 0:
            return None
        
        reportURL = reportURL[0]

        pageURL = urlparse(response.url)
        return "{}://{}{}".format(pageURL.scheme, pageURL.netloc, reportURL)
        # print('Response HTTP Status Code: {status_code}'.format(
            # status_code=response.status_code))
        # print('Response HTTP Response Body: {content}'.format(
            # content=response.content))
    except requests.exceptions.RequestException:
        # print('HTTP Request failed')
        return None




def csv_export(session, start_date, end_date):
    """
    Fetches a csv export report for the date range. Date params should be strings
    of the format mm/dd/yyy.
    """
    # CSV Export
    # POST https://carelink.minimed.com/patient/main/selectCSV.do

    try:
        response = session.post(
            url="https://carelink.minimed.com/patient/main/selectCSV.do",
            params={
                "t": "11",
            },
            headers={
                "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
                "User-Agent": user_agent,
            },
            data={
                "datePicker2": start_date,
                "listSeparator": ",",
                "datePicker1": end_date,
                "report": "11",
                "customerID": "553090",
            },
        )
        #sys.stderr.write('Response HTTP Response Body: {content}\n'.format(
         #   content=response.content))
        return response.content
    except requests.exceptions.RequestException:
        sys.stderr.write('HTTP Request failed\n')
        sys.stderr.write('Response HTTP Status Code: {status_code}\n'.format(
            status_code=response.status_code))


def sensor_24_hours(session):
    # Sensor Last 24 Hours
    # GET https://carelink.minimed.com/patient/connect/ConnectViewerServlet

    #sys.stderr.write("Fetching 24 hour sensor data\n")

    try:
        response = session.get(
            url="https://carelink.minimed.com/patient/connect/ConnectViewerServlet",
            params={
                "cpSerialNumber": "NONE",
                "msgType": "last24hours",
                "requestTime": time.time(),
            },
            headers={
                "User-Agent": user_agent,
            },
        )
        #sys.stderr.write('Response HTTP Status Code: {status_code}\n'.format(
          #  status_code=response.status_code))
        # sys.stderr.write('Response HTTP Response Body: {content}'.format(
            # content=response.content))
        return sensor_data_from_json(response.json())
    except ValueError:
        #sys.stderr.write("JSON decoding failed\n")
        return None
    except requests.exceptions.RequestException:
        #sys.stderr.write('HTTP Request failed\n')
        return None

def login(session, username, password):
    # Login
    # POST https://carelink.minimed.com/patient/j_security_check

    #sys.stderr.write("Logging In\n")
    try:
        response = session.post(
            url="https://carelink.minimed.com/patient/j_security_check",
            headers={
                "Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
                "User-Agent": user_agent
            },
            data={
                "j_password": password,
                "j_username": username,
                "j_character_encoding": "UTF-8",
            },
        )
        #sys.stderr.write('Response HTTP Status Code: {status_code}\n'.format(
        #    status_code=response.status_code))
        # sys.stderr.write('Response HTTP Response Body: {content}'.format(
            # content=response.content))
        return True
    except requests.exceptions.RequestException:
        #sys.stderr.write('HTTP Request failed\n')
        return False

def sensor_data_from_json(json):
    sd = SensorData()
    sd.lastSensorTimestampString = json["lastSensorTSAsString"]
    sd.sensorStateString = json["sensorState"]

    lastSG = json["lastSG"]
    if "sg" in lastSG and "datetime" in lastSG:
        sd.lastSensorReading = SensorReading(lastSG["sg"], lastSG["datetime"])

    insulin = json["activeInsulin"]
    if "amount" in insulin and "datetime" in insulin:
        sd.activeInsulin = Insulin(insulin["amount"], insulin["datetime"])

    for sg in json["sgs"]:
        if "sg" in sg and "datetime" in sg:
            sd.sensorReadings.append(SensorReading(sg["sg"], sg["datetime"]))
    
    return sd


class SensorReading(object):
    reading = 0.0
    dateTimeString = ""

    def __init__(self, r = 0.0, dt = ""):
        self.reading = r
        self.dateTimeString = dt

class Insulin(object):
    amount = 0.0
    dateTimeString = ""

    def __init__(self, a = 0.0, dt = ""):
        self.amount = a
        self.dateTimeString = dt

class SensorData(object):
    lastSensorTimestampString = ""
    sensorStateString = ""
    lastSensorReading = SensorReading()
    activeInsulin = Insulin()
    sensorReadings = []

    def __init__(self):
        pass

File: carelink.go
package main

import (
	"crypto/tls"
	"encoding/csv"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/Glucloser/models"
	"golang.org/x/net/publicsuffix"
	"io"
	"log"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/601.3.9 (KHTML, like Gecko)     Version/9.0.2 Safari/601.3.9"
	loginURL  = "https://carelink.minimed.com/patient/j_security_check"
	csvURL    = "https://carelink.minimed.com/patient/main/selectCSV.do?t=11"
	cgmURL    = "https://carelink.minimed.com/patient/connect/ConnectViewerServlet"
)

func ParseCSVExport(reader io.ReadCloser) (<-chan models.AuditItem, error) {
	csvReader := csv.NewReader(reader)
	csvReader.FieldsPerRecord = -1

	var headers []string
	for {
		line, err := csvReader.Read()
		if err != nil {
			return nil, err
		}
		if len(line) == 0 {
			continue
		}
		if line[0] == "Index" {
			headers = line
			break
		}
	}

	if len(headers) == 0 {
		return nil, errors.New("No headers in CSV")
	}

	out := make(chan models.AuditItem)
	go func(out chan models.AuditItem, csvReader *csv.Reader, closer io.Closer) {
		for {
			line, err := csvReader.Read()
			if err != nil {
				log.Printf("CSV read error %v", err)
				close(out)
				closer.Close()
				break
			}
			if len(line) == 0 {
				continue
			}
			item := &models.AuditItem{}
			for i := 0; i < len(headers) && i < len(line); i++ {
				item.SetRaw(headers[i], line[i])
			}
			out <- *item
		}
	}(out, csvReader, reader)

	return out, nil

}

func ParseCGMExport(reader io.ReadCloser) ([]models.Sugar, error) {
	decoder := json.NewDecoder(reader)
	var sgs struct {
		Sgs []struct {
			Value float64 `json:"sg"`
			Time  string  `json:"datetime"`
		} `json:"sgs"`
	}
	err := decoder.Decode(&sgs)
	if err != nil {
		return []models.Sugar{}, err
	}

	sugars := make([]models.Sugar, len(sgs.Sgs))
	for i, sg := range sgs.Sgs {
		sugar := models.Sugar{}
		occurredAt, err := time.Parse("Jan 2, 2006 15:04:05", sg.Time)
		if err != nil {
			continue
		}
		sugar.OccurredAt = occurredAt
		sugar.Value = int(sg.Value)
		sugars[i] = sugar
	}
	reader.Close()

	return sugars, nil

}

// CarelinkSession holds state for interacting with Carelink
type CarelinkSession struct {
	client http.Client
}

func NewCarelinkSession() (CarelinkSession, error) {
	defaultTransport := http.DefaultTransport.(*http.Transport)

	// Create new Transport that ignores self-signed SSL
	tr := &http.Transport{
		Proxy:                 defaultTransport.Proxy,
		DialContext:           defaultTransport.DialContext,
		MaxIdleConns:          defaultTransport.MaxIdleConns,
		IdleConnTimeout:       defaultTransport.IdleConnTimeout,
		ExpectContinueTimeout: defaultTransport.ExpectContinueTimeout,
		TLSHandshakeTimeout:   defaultTransport.TLSHandshakeTimeout,
		TLSClientConfig:       &tls.Config{InsecureSkipVerify: true},
	}
	jar, err := cookiejar.New(&cookiejar.Options{PublicSuffixList: publicsuffix.List})
	if err != nil {
		return CarelinkSession{}, err
	}
	return CarelinkSession{http.Client{Transport: tr, Jar: jar}}, nil
}

func (sess CarelinkSession) Login(username, password string) error {

	values := url.Values{}
	values.Add("j_username", username)
	values.Add("j_password", password)
	values.Add("j_character_encoding", "UTF-8")

	reader := strings.NewReader(values.Encode())

	req, err := http.NewRequest("POST", loginURL, reader)
	if err != nil {
		return err
	}

	req.Header.Add("User-Agent", userAgent)
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded; charset=utf-8")
	resp, err := sess.client.Do(req)
	if err != nil {
		return err
	}

	if resp.StatusCode > 400 {
		log.Printf("%s response from Carelink Login\n", resp.Status)
		return fmt.Errorf("%s response from Carelink Login", resp.Status)
	}

	return nil
}

func (sess CarelinkSession) CSVExport(start, end string) (io.ReadCloser, error) {
	values := url.Values{}
	values.Add("datePicker2", start)
	values.Add("listSeparator", ",")
	values.Add("datePicker1", end)
	values.Add("report", "11")
	values.Add("customerID", "553090")

	reader := strings.NewReader(values.Encode())

	req, err := http.NewRequest("POST", csvURL, reader)
	if err != nil {
		return nil, err
	}

	req.Header.Add("User-Agent", userAgent)
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded; charset=utf-8")
	resp, err := sess.client.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode > 400 {
		log.Printf("%s response from Carelink CSVExport\n", resp.Status)
		return nil, fmt.Errorf("%s response from Carelink CSVExport", resp.Status)
	}

	return resp.Body, nil
}

func (sess CarelinkSession) CGMExport() (io.ReadCloser, error) {
	urlParams := url.Values{}
	urlParams.Add("cpSerialNumber", "NONE")
	urlParams.Add("msgType", "last24hours")
	now := time.Now().Unix()
	urlParams.Add("requestTime", strconv.FormatInt(now, 10))
	url := cgmURL + "?" + urlParams.Encode()

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Add("User-Agent", userAgent)
	resp, err := sess.client.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode > 400 {
		log.Printf("%s response from Carelink CGMExport\n", resp.Status)
		return nil, fmt.Errorf("%s response from Carelink CGMExport", resp.Status)
	}

	return resp.Body, nil
}
